Correct module real path resolution

paztis opened the pull request:
- supports the monorepo correctly
- correct the no-extraneous-dependencies usage

paztis commented:
Not sure to understand why, but I have this CI build that fails only on node v10: https://ci.appveyor.com/project/benmosher/eslint-plugin-import/builds/31684756/job/7lv7l952s32a8qx7

It crash on resolving the .babelrc of the webpack resolver. I don't change this part.
I've try to use a real JSON for babelrc ({"extends": "../../.babelrc"}) but result is the same.

Understand the issue ?

Coverage decrease is because of the fallback on no-extraneous-dependencies we never pass in (it might never apperas in reality, but I keep it for security)

ljharb commented on package.json:
```diff
@@ -96,6 +96,7 @@
     "doctrine": "1.5.0",
     "eslint-import-resolver-node": "^0.3.2",
     "eslint-module-utils": "^2.4.1",
+    "find-up": "^2.0.0",
```
i would prefer not to depend on any modules from this author; they drop support for old nodes too aggressively.

paztis replied:
**find-up** is a dependency of **read-pkg-up**, so already present in nested deps. For performances I only need to get the package.json path, not to load its content, That's why I don't use **read-pkg-up** directly.
If needed I can use full local code instead of adding a dependency

ljharb replied:
ah, true, fair enough. i'll worry about excising those deps later.

ljharb commented on resolvers/webpack/.babelrc:
```diff
@@ -1 +1 @@
-../../.babelrc
\ No newline at end of file
+../../.babelrc
```
this was a symlink; why did this file change?

paztis replied:
this file import was failing on ApVeillor in case of node v10. 
I try to replace its content with 
```
{
  "extends": "../../.babelrc"
}
```
to avoid symling problem, but still present, so I rollback it.

I think the extension is a better way as the symlink for local development

paztis replied:
I don't know how to commit a sYmlinked file. can you do it after the merge ?

ljharb commented on src/core/importType.js:
```diff
@@ -1,6 +1,8 @@
+import * as PATH from 'path'
```
```suggestion
import path from 'path'
```
`import * as` is only for metaprogramming or laziness.

paztis replied:
Done. Was not aware Babel is differentiating es6 default export and commonjs module export here.
Classically I use webpack for this resolving (babel is not transpiling the imports) and webpack doesn't do this distinction.

PATH is preserved in uppercase to avoid collision will all **path** local var in the code.

ljharb commented on src/core/importType.js:
```diff
@@ -80,15 +84,16 @@ function isRelativeToSibling(name) {
 }
 
-function typeTest(name, settings, path) {
-  if (isAbsolute(name, settings, path)) { return 'absolute' }
```
please revert the unrelated changes here; the extra unused arguments are very intentional.

paztis replied:
done

ljharb commented on src/core/importType.js:
```diff
@@ -97,5 +102,5 @@ export function isScopedModule(name) {
 
 export default function resolveImportType(name, context) {
-  return typeTest(name, context.settings, resolve(name, context))
+  return typeTest(name, context.settings, resolve(name, context), context)
```
```suggestion
  return typeTest(name, context, resolve(name, context))
```
that way typeTest can read the settings off of the context object instead.

paztis replied:
done

ljharb commented on src/core/importType.js:
```diff
@@ -80,15 +84,16 @@ function isRelativeToSibling(name) {
-  if (isRelativeToParent(name, settings, path)) { return 'parent' }
-  if (isIndex(name, settings, path)) { return 'index' }
-  if (isRelativeToSibling(name, settings, path)) { return 'sibling' }
+function typeTest(name, settings, path, context) {
```
```suggestion
function typeTest(name, context, path) {
  const { settings } = context
```

paztis replied:
done

ljharb commented on src/core/packagePath.js:
```diff
@@ -0,0 +1,18 @@
+}
+
+export function getFilePackageName(filePath) {
+  const pkg = readPkgUp.sync({cwd: filePath, normalize: false}).pkg
```
```suggestion
  const { pkg } = readPkgUp.sync({cwd: filePath, normalize: false})
```

paztis replied:
done

paztis commented:
review changes applied

paztis commented:
Any new for the PR approval ?

paztis commented:
No news: start to be impossible to continue without thiese fixes
I will publish my custom version of eslint-plugin-import in case of no news :(

paztis commented:
any news ?

jimisaacs commented:
+1

geoffp commented:
Looking forward to this!

ljharb commented on src/rules/extensions.js:
```diff
@@ -130,7 +130,11 @@ module.exports = {
 
       if (isScopedModule(file) && slashCount <= 1) return true
-      if (isExternalModule(file, context, resolve(file, context)) && !slashCount) return true
+      if (isExternalModule(
+        file,
+        context.settings,
+        resolve(file, context),
+        context) && !slashCount) return true
```
```suggestion
      if (slashCount > 0 && isExternalModule(
        file,
        context.settings,
        resolve(file, context),
        context
      )) {
        return true
      }
```

ljharb commented on src/rules/extensions.js:
```diff
@@ -158,8 +162,11 @@ module.exports = {
       // determine if this is a module
-      const isPackage = isExternalModule(importPath, context.settings)
-        || isScoped(importPath)
+      const isPackage = isExternalModule(
+        importPath,
+        context.settings,
+        resolve(importPath, context),
+        context) || isScoped(importPath)
```
```suggestion
      const isPackage = isExternalModule(
        importPath,
        context.settings,
        resolve(importPath, context),
        context
      ) || isScoped(importPath)
```

ljharb commented on src/rules/no-extraneous-dependencies.js:
```diff
@@ -109,6 +110,17 @@ function optDepErrorMessage(packageName) {
 }
 
+function getModuleOriginalName(name) {
+  const splitName = name.split('/')
+  return splitName[0][0] === '@'
+    ? splitName.slice(0, 2).join('/')
+    : splitName[0]
```
```suggestion
  const [first, second] = name.split('/')
  return first.startsWith('@') ? `${first}/${second}` : first
```

ljharb commented on tests/src/core/resolve.js:
```diff
@@ -99,13 +99,14 @@ describe('resolve', function () {
-                      , Object.assign({}, testContext, { getFilename: function () { return utils.getFilename('foo.js') } }),
-                      )).to.equal(utils.testFilePath('./bar.jsx'))
-  })
+  // crash because context.report doesn't exist
```
why would `context.report` not exist?

paztis replied:
have you try to run the tests ? context was always null for me on windows

ljharb replied:
That doesn't make any sense; `context.report` is the way eslint creates all of its warnings.

ljharb commented on tests/src/rules/order.js:
```diff
@@ -325,21 +325,21 @@ ruleTester.run('order', rule, {
-        'import/external-module-folders': ['node_modules', 'files/symlinked-module'],
-      },
-    }),
+    // test({
+    //   code: `
+    //     import _ from 'lodash';
+    //     import m from '@test-scope/some-module';
+    //
+    //     import bar from './bar';
+    //   `,
+    //   options: [{
+    //     'newlines-between': 'always',
+    //   }],
+    //   settings: {
+    //     'import/resolver': 'webpack',
+    //     'import/external-module-folders': ['node_modules', 'files/symlinked-module'],
+    //   },
+    // }),
```
why is this commented out?

paztis replied:
files/symlinked-module is a bad folder: root of the package here it tests/files/ folder because of a package.json present in this file.
For the module resolution it results in a folder tests/file/files/symlinked-module folder that didn't exist.
It was working in the past because module resolution was not real one like mine

ljharb commented on tests/src/rules/order.js:
```diff
@@ -353,7 +353,7 @@ ruleTester.run('order', rule, {
       settings: {
         'import/resolver': 'node',
-        'import/external-module-folders': ['node_modules', 'files/symlinked-module'],
+        'import/external-module-folders': ['node_modules', 'symlinked-module'],
```
why does this not need the `files/` prefix?

paztis commented:
Hi

I review all the changes.

I've only problems with the babelrc file: I cannot commit the original symlink (don't know how to do it). Can you fix it on your side after the merge ?

For the **files/symlinked-module** change, it is because my module resolution is now really looking at the root package path + folder name, and not only test if the folder name is in the path.
I your unit test, as there's a package.json under the folder **tests/src/files**, it is considered as the root, and not tests **folder** as you imagine.
Forced to remove it.
Potentially you may reorganize all the test files under correct folder structure, with correct package.json like in reality.


Please tell me if there's missing things

SanjiKir commented:
Hey guys, will this PR resolve https://github.com/benmosher/eslint-plugin-import/issues/1832? I have the same problem with `import/named` in monorepo structure. It doesn't work for yarn workspaces.

paztis commented:
No, I think. It only verify the import path is correctly written. Not it the content of the import is good.
We pack can do it potentially. Or typescript

SanjiKir commented:
@paztis Yeah we are migrating to TS so far, so it will take some time since the project is huge. I didn't get the last sentence of yours 
`We pack can do it potentially.` What do you mean by that?

paztis commented:
Sorry autocorrect from my phone.
We can do it with webpack or typescript I said.

In webpack there's an option to check the name import really exist in the
import file

Le ven. 31 juil. 2020 à 12:08, Shamanin Kirill <notifications@github.com> a
écrit :

> @paztis <https://github.com/paztis> Yeah we are migrating to TS so far,
> so it will take some time since the project is huge. I didn't get the last
> sentence of yours
> We pack can do it potentially. What do you mean by that?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/benmosher/eslint-plugin-import/pull/1696#issuecomment-667045168>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABIKMMNC26KBLPJHF4BTFJLR6KJZBANCNFSM4LSVAZOA>
> .
>

paztis commented:
@ljharb what is the review status now ?

caseyWebb commented:
@paztis I found your fork on npm, but how can I actually use it?

I added `"plugins": ["@paztis/eslint-plugin-import"]` to my .eslintrc, but I can't figure out how to configure rules. e.g. `"import/order"`, `"@pastiz/eslint-plugin-import/order"`, and half a dozen other combinations all throw "Definition for rule was not found" errors.

ljharb commented:
@paztis would you mind rebasing this on top of latest master (which will remove all merge commits) and ideally condense the commits? There's 34 right now.

I can handle https://github.com/benmosher/eslint-plugin-import/pull/1696#discussion_r461985866 for you after that's done, and give it a fresh review.

paztis commented:
In fact the problem is Airbnb is pointing to original eslint-plugin-import,
which conflict with mine .
That's why I push to merge my changes.

To use it, the only way I found is to use symlinks in my local environment:
I symlink original one to mine.

Le mar. 25 août 2020 à 08:47, Jordan Harband <notifications@github.com> a
écrit :

> @paztis <https://github.com/paztis> would you mind rebasing this on top
> of latest master (which will remove all merge commits) and ideally condense
> the commits? There's 34 right now.
>
> I can handle #1696 (comment)
> <https://github.com/benmosher/eslint-plugin-import/pull/1696#discussion_r461985866>
> for you after that's done, and give it a fresh review.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/benmosher/eslint-plugin-import/pull/1696#issuecomment-679837537>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABIKMMJCMELSI66K27ETSO3SCNM7NANCNFSM4LSVAZOA>
> .
>

paztis commented:
@ljharb rebase on top of master is done not
you can do the merge ?

michaelfarrell76 commented:
so excited for this! thank you all so much 🙏

paztis commented:
Me too ^^
@ljharb any news for the merge ?

ljharb commented:
@paztis thanks! I've rebased this PR.

I've split out the `isExternalModule` bugfix into its own commit. That commit includes a test for `order` that fails without the fix, but i can't come up with a test for `extensions` that fails without the fix. Got any ideas?

Also, the second commit is quite large, so it's hard to me to understand everything it's doing. Can any parts of it be split into separate commits, that could land "first" - before the primary fix in the PR?

paztis commented:
For the extension, you were testing only the name not the real path.
If an alias exist for this name, potentially it is not an external.
Is package value was potentially incorrect, then
isUseOfExtensionRequired(extension, package)
Will fail

I just fix the usage of is external module because the last parameters was
missing.

For the second commit, tell me which point you want to check.

Le sam. 5 sept. 2020 à 23:18, Jordan Harband <notifications@github.com> a
écrit :

> @paztis <https://github.com/paztis> thanks! I've rebased this PR.
>
> I've split out the isExternalModule bugfix into its own commit. That
> commit includes a test for order that fails without the fix, but i can't
> come up with a test for extensions that fails without the fix. Got any
> ideas?
>
> Also, the second commit is quite large, so it's hard to me to understand
> everything it's doing. Can any parts of it be split into separate commits,
> that could land "first" - before the primary fix in the PR?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/benmosher/eslint-plugin-import/pull/1696#issuecomment-687664431>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABIKMMPAHEG7762HLSNGV3DSEKTLBANCNFSM4LSVAZOA>
> .
>

paztis commented:
@ljharb, any news ?

buchanaf commented:
Hey guys (@ljharb and @paztis ), really looking forward to having this support. Is there anything I could do to help speed this along so this PR doesn't get lost in purgatory?

paztis commented:
I don't know in fact. I'm still waiting the approval for the merge.

BPreisner commented:
@paztis Could you please look at the failed build, I think that @ljharb is waiting for your action ;)

michaelfarrell76 commented:
thanks for merging this! @ljharb any idea when a new version will be published to npm?
